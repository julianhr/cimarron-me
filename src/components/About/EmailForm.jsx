import React from 'react'
import PropTypes from 'prop-types'
import { css } from '@emotion/core'

import Form from '~/library/forms/Form'
import Input from '~/library/forms/Input'
import Textarea from '~/library/forms/Textarea'


function EmailForm({
  SubmitButton,
  defaultValues,
  fieldErrors,
  formRefs,
  onFieldChange,
  onSubmit,
}) {
  return (
    <Form
      formStyle={css`
        align-self: center;
        width: 80%;
        max-width: 560px;
      `}
      onSubmit={onSubmit}
    >
      <Input
        fwdRef={formRefs.email}
        defaultValue={defaultValues.email}
        errorTxt={fieldErrors.email}
        labelTxt='Your Email'
        name='email'
        type='email'
        onChange={onFieldChange}
      />
      <Input
        fwdRef={formRefs.name}
        defaultValue={defaultValues.name}
        errorTxt={fieldErrors.name}
        labelTxt='Your Name'
        name='name'
        onChange={onFieldChange}
      />
      <Input
        fwdRef={formRefs.subject}
        defaultValue={defaultValues.subject}
        errorTxt={fieldErrors.subject}
        labelTxt='Subject'
        name='subject'
        onChange={onFieldChange}
      />
      <Textarea
        fwdRef={formRefs.message}
        defaultValue={defaultValues.message}
        errorTxt={fieldErrors.message}
        labelTxt='Message'
        name='message'
        onChange={onFieldChange}
      />
      <SubmitButton />
    </Form>
  )
}

EmailForm.propTypes = {
  defaultValues: PropTypes.object.isRequired,
  fieldErrors: PropTypes.object.isRequired,
  formRefs: PropTypes.object.isRequired,
  onFieldChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  SubmitButton: PropTypes.func.isRequired,
}

export default EmailForm
