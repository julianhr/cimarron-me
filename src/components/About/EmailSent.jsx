import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'

import imgSrc from '~/assets/img/thumbs_up.png'
import ImageLoader from '~/library/ImageLoader'


const Root = styled.div`
  align-self: center;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 80px 0 0;

  h3 {
    color: ${props => props.theme.colors.text};
  }
`


function EmailSent() {
  return (
    <Root>
      <ImageLoader
        maxWidth={339}
        maxHeight={364}
        imgSrc={imgSrc}
        shouldFadeIn={false}
      />
      <h3
        css={{ marginBottom: 0 }}
      >
        Message successfully sent
      </h3>
      <h3>Thank you!</h3>
    </Root>
  )
}

export default EmailSent
