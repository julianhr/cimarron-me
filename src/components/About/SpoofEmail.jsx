import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'

import SvgImage from '~/assets/img/suspicious_emoji.svg'


const Root = styled.div`
  align-self: center;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 80px 0 0;
  text-align: center;

  h2 {
    color: ${props => props.theme.colors.text};
  }
`


function EmailSent() {
  return (
    <Root>
      <h2
        css={{ marginBottom: 80 }}
      >
        That was an awefully suspicious email
      </h2>
      <SvgImage
        width={240}
        height={240}
      />
      <h2
        css={{ marginTop: 80 }}
      >
        I have reported it to my detective minions
      </h2>
    </Root>
  )
}

export default EmailSent
