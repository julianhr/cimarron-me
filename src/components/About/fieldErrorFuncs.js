import emailValidator from 'email-validator'

export const ERRORS = {
  required: () => 'This field is required',
  length: (min, max) => `This field must be between ${min} and ${max} characters`,
  email: () => 'This email address is invalid',
}


const fieldErrorFuncs = {
  email: (value) => {
    if (!value) {
      return ERRORS.required()
    } else if (value.length < 8 || value.length > 50) {
      return ERRORS.length(8, 50)
    } else if (!emailValidator.validate(value)) {
      return ERRORS.email()
    }
  },
  name: (value) => {
    if (!value) {
      return ERRORS.required()
    } else if (value.length < 2 || value.length > 50) {
      return ERRORS.length(2, 50)
    }
  },
  subject: (value) => {
    if (!value) {
      return ERRORS.required()
    } else if (value.length < 2 || value.length > 120) {
      return ERRORS.length(2, 120)
    }
  },
  message: (value) => {
    if (!value) {
      return ERRORS.required()
    } else if (value.length < 50 || value.length > 10000) {
      return ERRORS.length(50, '10,000')
    }
  },
}

export default fieldErrorFuncs
