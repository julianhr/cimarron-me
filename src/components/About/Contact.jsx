import React from 'react'
import styled from '@emotion/styled'
import { debounce } from 'lodash-es'

import { urlBuilder } from '~/utils'
import EmailForm from './EmailForm'
import EmailSent from './EmailSent'
import SpoofEmail from './SpoofEmail'
import Button from '~/library/Button'
import fieldErrorFuncs from './fieldErrorFuncs'


const H3 = styled.h3`
  margin-top: 100px;
  align-self: center;
  color: ${props => props.theme.colors.text};
`

const ErrorMessage = styled.div`
  align-self: center;
  color: ${props => props.theme.colors.error};
  max-width: 560px;
`

class Contact extends React.PureComponent {
  constructor() {
    super()
    this.handleFieldOnChange = this.handleFieldOnChange.bind(this)
    this.handleFormOnSubmit = this.handleFormOnSubmit.bind(this)
    this.debFieldOnChange = this.debFieldOnChange.bind(this)
  }

  state = {
    wasFormSubmitted: false,
    emailStatus: 'draft',
    fieldValues: {},
    fieldErrors: {},
    errorMessage: '',
  }

  formRefs = {
    // form
    submit: React.createRef(),
    // fields
    email: React.createRef(),
    name: React.createRef(),
    subject: React.createRef(),
    message: React.createRef(),
  }

  componentDidMount() {
    this.setCsrfToken()
  }

  handleFieldOnChange({ target: { name, value } }) {
    this.debFieldOnChange.call(this, name, value)
  }

  debFieldOnChange = debounce((name, value) => {
    const { fieldValues, fieldErrors, wasFormSubmitted }  = this.state
    const nextState = { fieldValues: { ...fieldValues, [name]: value } }

    if (wasFormSubmitted) {
      const fieldError = fieldErrorFuncs[name].call(this, value)

      if (fieldError) {
        nextState.fieldErrors = { ...fieldErrors, [name]: fieldError }
      } else {
        if (fieldErrors[name]) {
          const nextFieldErrors = {...fieldErrors}
          delete nextFieldErrors[name]
          nextState.fieldErrors = nextFieldErrors
        }
      }
    }

    this.setState(nextState, () => {
      wasFormSubmitted && this.formRefs[name].current.focus()
    })
  }, 150)

  async setCsrfToken() {
    const urlCsrfToken = urlBuilder('/cimarron/generate-token')
    const res = await fetch(urlCsrfToken)

    if (res.ok) {
      const json = await res.json()
      this.setState({ fieldValues: { ...this.state.fieldValues, csrf_token: json.token } })
    }
  }

  async handleFormOnSubmit(event) {
    event.preventDefault()
    this.formRefs.submit.current.disabled = true

    if (!this.state.wasFormSubmitted) {
      this.setState({ wasFormSubmitted: true })
    }

    try {
      await this.setFieldErrors()
      const resp = await this.sendEmail()
      await this.checkSendEmailRespErrors(resp)
      this.setState({ emailStatus: 'sent' })
    } catch (error) {
      null
    }

    if (this.formRefs.submit.current) {
      this.formRefs.submit.current.disabled = false
    }
  }

  async setFieldErrors() {
    const fieldErrors = {}

    for (let [name, fn] of Object.entries(fieldErrorFuncs)) {
      const error = fn(this.state.fieldValues[name])
      if (error) { fieldErrors[name] = error }
    }

    this.setState({ fieldErrors })
    const hasErrors = Object.keys(fieldErrors).length > 0
    if (hasErrors) { throw new Error('fields have errors') }
  }

  async sendEmail() {
    try {
      const url = urlBuilder('/cimarron/email')
      const resp = await fetch(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(this.state.fieldValues),
      })
      return resp
    } catch (error) {
      this.setState({ emailStatus: 'error', errorMessage: error.message })
      throw error
    }
  }

  async checkSendEmailRespErrors(resp) {
    if (resp.ok) { return }

    let nextState

    if (('' + resp.status)[0] === '5') {
      // 5xx errors
      const errorText = await resp.text()
      nextState = { emailStatus: 'error', errorMessage: errorText }
    } else {
      // 4xx errors
      const json = await resp.json()

      if (json.error_type === 'ValidationError' && json.error_payload.csrf_token) {
        nextState = { emailStatus: 'spoof', errorMessage: '' }
      } else if (json.error_type === 'ValidationError') {
        nextState = { emailStatus: 'invalid_data', fieldErrors: json.error_payload }
      } else {
        const errorText = await resp.text()
        nextState = { emailStatus: 'error', errorMessage: errorText }
      }
    }

    this.setState(nextState)
    throw new Error('response error')
  }

  getEmailFormBtn = () => (
    () => (
      <Button
        fwdRef={this.formRefs.submit}
        type={'submit'}
        btnStyle={{ alignSelf: 'flex-end'}}
      >
        Send
      </Button>
    )
  )

  renderErrorFlash() {
    if (this.state.emailStatus !== 'error') { return }

    return (
      <ErrorMessage>
        <p>
          There was an error sending this email:
          <br />
          {this.state.errorMessage}
        </p>
      </ErrorMessage>
    )
  }

  render() {
    if (!this.state.fieldValues.csrf_token) {
      return null
    }

    switch (this.state.emailStatus) {
      case 'draft':
      case 'error':
      case 'invalid_data':
        return (
          <>
            <H3>Contact</H3>
            <EmailForm
              defaultValues={this.state.fieldValues}
              fieldErrors={this.state.fieldErrors}
              formRefs={this.formRefs}
              onFieldChange={this.handleFieldOnChange}
              onSubmit={this.handleFormOnSubmit}
              SubmitButton={this.getEmailFormBtn()}
            />
            {this.renderErrorFlash()}
          </>
        )
      case 'sent':
        return <EmailSent />
      case 'spoof':
        return <SpoofEmail />
    }
  }
}

export default Contact
