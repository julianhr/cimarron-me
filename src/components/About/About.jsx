import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'

import ErrorBoundary from '~/library/ErrorBoundary'
import AnchorTag from '~/library/AnchorTag'
import ImageLoader from '~/library/ImageLoader'
import linkedInBtnSrc from '~/assets/img/linkedin_btn.png'
import Contact from './Contact'


const Root = styled.section`
  display: flex;
  flex-direction: column;
`

const H2 = styled.h2`
  align-self: center;
`

function About() {
  return (
    <Root>
      <H2>About</H2>
      <AnchorTag
        url='https://www.linkedin.com/in/julianhr/'
        isLinkRouted={false}
        isTargetBlank
        rootStyle={{ display: 'contents' }}
      >
        <ImageLoader
          maxWidth={200}
          maxHeight={37}
          imgSrc={linkedInBtnSrc}
        />
      </AnchorTag>
      <ErrorBoundary>
        <Contact />
      </ErrorBoundary>
    </Root>
  )
}

export default About
