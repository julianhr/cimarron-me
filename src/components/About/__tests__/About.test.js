import React from 'react'
import { render, mount, shallow } from 'enzyme'
import { create } from 'react-test-renderer'

import About from '../About'
import MockApp from '~/__tests__/__mocks__/MockApp'


describe('About', () => {
  it('matches snapshot', () => {
    const wrapper = create(
      <MockApp>
        <About />
      </MockApp>
    )

    expect(wrapper).toMatchSnapshot()
  })
})

