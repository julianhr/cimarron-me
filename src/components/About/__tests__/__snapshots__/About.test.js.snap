// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`About matches snapshot 1`] = `
<section
  className="css-14jo2cv-Root e5wwmhd0"
>
  <h2
    className="css-1s6s8zy-H2 e5wwmhd1"
  >
    About
  </h2>
  <a
    className="css-w4rclr-AnchorTag"
    href="https://www.linkedin.com/in/julianhr/"
    rel="noopener noreferrer"
    target="_blank"
  >
    <figure
      className="css-1uf6ek7-Root-ImageLoader eeswxza0"
    >
      <img
        className="css-1k90mm3"
        onLoad={[Function]}
        src="test-file-stub"
      />
    </figure>
  </a>
</section>
`;
